File: src/main/java/com/github/nsc/de/shake/lexer/characterinputstream/position/PositionMaker.java
package com.github.nsc.de.shake.lexer.characterinputstream.position;

import com.github.nsc.de.shake.lexer.characterinputstream.charactersource.CharacterSource;

import java.util.ArrayList;
import java.util.List;

/**
 * A modifiable {@link PositionMarker} that stores the positions
 * of the line-separators
 */
public class PositionMaker implements PositionMarker {

    /**
     * The index of the position
     */
    private int index;

    /**
     * The column of the position
     */
    private int column;

    /**
     * The line of the position
     */
    private int line;

    /**
     * Stores the locations of the line-separators
     */
    private List<Integer> lineSeparators = new ArrayList<>();

    /**
     * The source of the {@link PositionMaker}
     */
    private final CharacterSource source;

    /**
     * Constructor for the {@link PositionMaker}
     *
     * @param index The {@link PositionMaker#index} of the position
     * @param column The {@link PositionMaker#column} of the position
     * @param line The {@link PositionMaker#line} of the position
     * @param source the {@link CharacterSource} the chars come from
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    public PositionMaker(int index, int column, int line, CharacterSource source) {
        this.index = index;
        this.column = column;
        this.line = line;
        this.source = source;
    }

    /**
     * Constructor for the {@link PositionMaker}
     *
     * @param source the {@link CharacterSource} the chars come from
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    public PositionMaker(CharacterSource source) {
        this.source = source;
        this.index = -1;
        this.column = 0;
        this.line = 1;
    }


    // ***********************************************
    // Getters

    /**
     * Getter for {@link PositionMaker#index}
     *
     * @return The {@link PositionMaker#index} of the {@link PositionMaker}
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    @Override
    public int getIndex() {
        return this.index;
    }

    /**
     * Getter for {@link PositionMaker#column}
     *
     * @return The {@link PositionMaker#column} of the {@link PositionMaker}
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    @Override
    public int getColumn() {
        return this.column;
    }

    /**
     * Getter for {@link PositionMaker#line}
     *
     * @return The {@link PositionMaker#line} of the {@link PositionMaker}
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    @Override
    public int getLine() {
        return this.line;
    }

    /**
     * Getter for {@link PositionMaker#source}
     *
     * @return the {@link PositionMaker#source} of the {@link PositionMaker}
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    public CharacterSource getSource() {
        return source;
    }

    public int[] getLineSeparators() {
        // Convert map to int[]
        int[] arr = new int[this.lineSeparators.size()];
        for(int i = 0; i < arr.length; i++) arr[i] = this.lineSeparators.get(i);
        return arr;
    }



    // ***********************************************
    // Others

    /**
     * Increases the {@link PositionMaker#index} and {@link PositionMaker#column}
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    public void nextColumn() {
        this.index++;
        this.column++;
    }

    /**
     * Increases the {@link PositionMaker#index} and {@link PositionMaker#line} and sets the
     * {@link PositionMaker#column} to 1
     *
     * @author <a href="https://github.com/nsc-de">Nicolas Schmidt &lt;@nsc-de&gt;</a>
     */
    public void nextLine() {
        this.lineSeparators.add(++this.index);
        this.line++;
        this.column = 1;
    }

    public PositionMap createPositionMap() {
        return new PositionMap(this.getSource(), this.getLineSeparators());
    }

    public Position createPositionAtLocation() {
        return new Position(createPositionMap(), getIndex(), getColumn(), getLine());
    }
}

File: util/src/commonMain/kotlin/com/github/shakelang/shake/util/characterinput/position/PositionMaker.kt
package com.github.shakelang.shake.util.characterinput.position

import com.github.shakelang.shake.util.characterinput.charactersource.CharacterSource

/**
 * A modifiable [PositionMarker] that stores the positions
 * of the line-separators
 */
@Suppress("unused")
class PositionMaker : PositionMarker {
    /**
     * Getter for [PositionMaker.index]
     *
     * @return The [PositionMaker.index] of the [PositionMaker]
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    /**
     * The index of the position
     */
    override var index: Int
        private set
    /**
     * Getter for [PositionMaker.column]
     *
     * @return The [PositionMaker.column] of the [PositionMaker]
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    /**
     * The column of the position
     */
    override var column: Int
        private set
    /**
     * Getter for [PositionMaker.line]
     *
     * @return The [PositionMaker.line] of the [PositionMaker]
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    /**
     * The line of the position
     */
    override var line: Int
        private set

    /**
     * Stores the locations of the line-separators
     */
    private val lineSeparators: MutableList<Int> = ArrayList()
    /**
     * Getter for [PositionMaker.source]
     *
     * @return the [PositionMaker.source] of the [PositionMaker]
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    /**
     * The source of the [PositionMaker]
     */
    val source: CharacterSource

    /**
     * Constructor for the [PositionMaker]
     *
     * @param index The [PositionMaker.index] of the position
     * @param column The [PositionMaker.column] of the position
     * @param line The [PositionMaker.line] of the position
     * @param source the [CharacterSource] the chars come from
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    constructor(index: Int, column: Int, line: Int, source: CharacterSource) {
        this.index = index
        this.column = column
        this.line = line
        this.source = source
    }

    /**
     * Constructor for the [PositionMaker]
     *
     * @param source the [CharacterSource] the chars come from
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    constructor(source: CharacterSource) {
        this.source = source
        index = -1
        column = 0
        line = 1
    }

    // ***********************************************
    // Getters
    fun getLineSeparators(): IntArray {
        // Convert map to int[]
        val arr = IntArray(lineSeparators.size)
        for (i in arr.indices) arr[i] = lineSeparators[i]
        return arr
    }
    // ***********************************************
    // Others
    /**
     * Increases the [PositionMaker.index] and [PositionMaker.column]
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    fun nextColumn() {
        index++
        column++
    }

    /**
     * Increases the [PositionMaker.index] and [PositionMaker.line] and sets the
     * [PositionMaker.column] to 1
     *
     * @author [Nicolas Schmidt &lt;@nsc-de&gt;](https://github.com/nsc-de)
     */
    fun nextLine() {
        lineSeparators.add(++index)
        line++
        column = 1
    }

    fun createPositionMap(): PositionMap {
        return PositionMap(source, getLineSeparators())
    }

    fun createPositionAtLocation(): Position {
        return Position(createPositionMap(), index, column, line)
    }
}